### PR TITLE
Build docs in a container

### PR DESCRIPTION
**Pull request by Lex-2008**
Disclaimer: `install.sh` and `main.sh` scripts are mostly copy-pasted form Jenkins config, with minor modifications like:

* not installing stuff required for building CFEngine
* unpacking CFEngine *.deb file instead of installing it

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a pull request! Maybe @olehermanse can review this?

**Comment by Lex-2008** [comment]
@vpodzime I can't request you in reviewers, so pinging here

File: build/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM ubuntu:16.04
+
+# We need sudo before all
+RUN apt-get -qy update && \
+    apt-get -qy install sudo
+
+# Add jenkins user
+RUN useradd -s /bin/bash jenkins
+RUN echo "jenkins    ALL=(ALL)       NOPASSWD: ALL">> /etc/sudoers
```
Comment (nickanderson) [inline build/Dockerfile:9]: this violates the security policy

**Review by nickanderson — APPROVED** [review]

File: build/Dockerfile (hunk 1 above)
Comment (vpodzime) [reply build/Dockerfile:9]: Yes, but it's better than running all commands as `root` even inside the container.

File: build/install.sh [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+# sudo pip3 install weasyprint
+# weasyprint --help
+# will it work on second try?
+# sudo pip3 install weasyprint
+# weasyprint --help
+
+# install wkhtmltopdf for creating PDFs
+# sudo apt-get -y install xfonts-base xfonts-75dpi fonts-liberation
+# wget https://downloads.wkhtmltopdf.org/0.12/0.12.5/wkhtmltox_0.12.5-1.bionic_amd64.deb
+# sudo dpkg -i wkhtmltox_0.12.5-1.bionic_amd64.deb
```
Comment (vpodzime) [inline build/install.sh:52]: Remove the above commented-out lines or put them into some `========` section with an explanation of why we want to keep them here.

File: build/install.sh [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+## Install RVM and Ruby
+## RVM helps us manage the ruby dependencies and ruby versions in a contained way.
+
+# Install mpapis public key (might need `gpg2` and or `sudo`)
+#keyserver=keys.gnupg.net
+keyserver=keyserver.ubuntu.com
+gpg --keyserver hkp://$keyserver --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
+gpg2 --keyserver hkp://$keyserver --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
```
Comment (vpodzime) [inline build/install.sh:61]: One `gpg` should be enough. Also, can we use `hkps://`?

File: build/install.sh [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+echo "Downloading and installing Ruby Version Manager"
+# Download the installer
+\curl -O https://raw.githubusercontent.com/rvm/rvm/master/binscripts/rvm-installer
+\curl -O https://raw.githubusercontent.com/rvm/rvm/master/binscripts/rvm-installer.asc
```
Comment (vpodzime) [inline build/install.sh:66]: Why the backslashes?

File: build/main.sh [hunk 5]
```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+
```
Comment (vpodzime) [inline build/main.sh:2]: Please add a check of `$#`

File: build/main.sh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+export FLAG_FILE_URL="http://buildcache.cfengine.com/packages/$PACKAGE_JOB/$PACKAGE_UPLOAD_DIRECTORY/PACKAGES_HUB_x86_64_linux_ubuntu_16/core-commitID"
+export NO_OUTPUT_DIR=1
+
+env
+set -x
+
+# take ownersip of all files
+sudo chown -R jenkins:jenkins .
+
+export WRKDIR=`pwd`
```
Comment (vpodzime) [inline build/main.sh:18]: Why `export`?

File: build/main.sh [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+### download CFEngine ###
+
+# c https://github.com/cfengine/misc/blob/master/vagrant_quickstart/build.sh
+
+function fetch_file() {
+  # $1 -- URL to fetch
+  # $2 -- destination
+  # $3 -- number of tries (with 10s pauses) [optional, default=1]
+  target="$1"
+  destination="$2"
```
Comment (vpodzime) [inline build/main.sh:32]: Should be `local`

File: build/main.sh [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+sudo chown -R jenkins:jenkins .
+
+export WRKDIR=`pwd`
+
+cd $WRKDIR/documentation-generator
+
+
+### download CFEngine ###
+
+# c https://github.com/cfengine/misc/blob/master/vagrant_quickstart/build.sh
```
Comment (vpodzime) [inline build/main.sh:25]: ?

File: build/main.sh [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+function fetch_file() {
+  # $1 -- URL to fetch
+  # $2 -- destination
+  # $3 -- number of tries (with 10s pauses) [optional, default=1]
+  target="$1"
+  destination="$2"
+  if [ $# -gt 2 ]; then
+    tries="$3"
+  else
+    tries=1
```
Comment (vpodzime) [inline build/main.sh:36]: And this too.

File: build/run.sh [hunk 10]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
```
Comment (vpodzime) [inline build/run.sh:2]: You can check the required variables here just like we do it in https://github.com/cfengine/masterfiles/blob/master/templates/federated_reporting/import.sh#L12

**Review by vpodzime — COMMENTED** [review]
Looks generally good to me. But we should either polish things a bit or create a follow-up ticket to do so. Now that this is reviewed and not just part of jenkins config XMLs, it requires a bit more cleanliness and shine.

File: build/main.sh (hunk 6 above)
Comment (Lex-2008) [reply build/main.sh:18]: later it's used in _run_jekyll.sh script, which expects it to be set to directory where all repos are cloned (and defaults to `pwd` also). But main.sh cd's into `documentation-generator` subdir somewhere before running _run_jekyll.sh.

It's possible to remove `export` here, and `cd -` before running _run_jekyll.sh.

Or change _run_jekyll.sh to set WRKDIR to "`pwd`/..".

Which of these options sounds less terrible? :)

File: build/main.sh (hunk 8 above)
Comment (Lex-2008) [reply build/main.sh:25]: where this function was copied from?

Or a reminder to share improvements with that script, too:
https://github.com/cfengine/misc/pull/146

File: build/main.sh (hunk 6 above)
Comment (vpodzime) [reply build/main.sh:27]: `_run_jekyll.sh` should use something like `"$(dirname $0)/../"`or whatever exactly it needs.

**Review by vpodzime — APPROVED** [review]
Looks good to me. Is there anything I need to do on *oslo-dc* for this, @Lex-2008 ?

**Comment by Lex-2008** [comment]
no. Containers work everywhere!

But I'd like to fix docs before merging this :)

**Pull request merged**